pkg/applicationserver/last_seen.go

```diff
@@ -0,0 +1,180 @@
+// Copyright © 2022 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package applicationserver
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+	"go.thethings.network/lorawan-stack/v3/pkg/applicationserver/metadata"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/unique"
+)
+
+type LastSeenProvider struct {
+	ctx     context.Context
+	cluster metadata.ClusterPeerAccess
+	ticker  <-chan time.Time
+
+	mapCapacity int
+	lastSeenMap map[string]time.Time
+	lsMu        sync.Mutex
+}
+
+// StoreInMap stores the timestamp in map if its younger than the current timestamp.
+// If map is full it doesn't store the new entry and returns false.
+func (ls *LastSeenProvider) StoreInMap(ids string, lastSeen *types.Timestamp) ([]*ttnpb.EndDevice, error) {
```

Comment:
I think that:
- `ids` should be `uid`, and really be a unique ID.
- The return should be a `map[string]*ttnpb.EndDevice` in order to allow for easier partitioning in Enterprise.